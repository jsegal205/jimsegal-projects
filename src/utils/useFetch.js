import { useState, useEffect } from "react";
import axios from "axios";

// custom hook for performing GET request
const useFetch = (url, initialValue) => {
  const [data, setData] = useState(initialValue);
  const [loading, setLoading] = useState(true);
  useEffect(() => {
    const fetchData = async () => {
      try {
        setLoading(true);
        const response = await axios.get(url);
        setData(response.data);
      } catch ({ response }) {
        setData({ status: response.status });
      } finally {
        setLoading(false);
      }
    };
    fetchData();
  }, [url]);
  return { loading, data };
};

export default useFetch;
