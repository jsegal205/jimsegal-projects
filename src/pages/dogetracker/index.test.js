import React from "react";
import { BrowserRouter as Router } from "react-router-dom";

import useFetch from "../../utils/useFetch";
import DogeTracker from "./index";
import { render } from "@testing-library/react";

jest.mock("../../utils/useFetch", () => jest.fn());

describe("DogeTracker Component", () => {
  describe("when `useFetch` is awaiting promise to resolve", () => {
    it("displays `Loading...`", () => {
      useFetch.mockReturnValue({
        loading: true,
        data: null,
      });

      const { container } = render(
        <Router>
          <DogeTracker />
        </Router>
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe("when `useFetch` returns data", () => {
    it("displays data", () => {
      useFetch.mockReturnValue({
        loading: false,
        data: {
          market_data: {
            current_price: {
              usd: 100,
              btc: 1,
            },
          },
        },
      });

      const { container } = render(
        <Router>
          <DogeTracker />
        </Router>
      );
      expect(container).toMatchSnapshot();
    });

    describe("when 500 status returned", () => {
      it("displays error messaging", () => {
        useFetch.mockReturnValue({
          loading: false,
          data: {
            error: "server likely on fire",
            status: 500,
          },
        });

        const { container } = render(
          <Router>
            <DogeTracker />
          </Router>
        );
        expect(container).toMatchSnapshot();
      });
    });
  });
});
