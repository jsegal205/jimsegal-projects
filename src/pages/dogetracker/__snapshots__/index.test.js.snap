// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DogeTracker Component when \`useFetch\` is awaiting promise to resolve displays \`Loading...\` 1`] = `
<div>
  <section
    class="loader"
  >
    <div
      class="loader-wheel"
    />
    <div
      class="loader-text"
    >
      Loading
    </div>
  </section>
</div>
`;

exports[`DogeTracker Component when \`useFetch\` returns data displays data 1`] = `
<div>
  <section
    class="doge-cointainer"
  >
    <h2
      class="doge-title"
    >
      What is Dogecoin's current price?
    </h2>
    <section
      class="doge-cointainer"
    >
      <img
        alt="a spinning coin with with the official dogecoin logo on one side and a shiba inu with the word WOW embossed on the coin"
        class="doge-image"
        src="https://i.imgur.com/jEll6Tr.gif"
      />
      <h3
        class="doge-price"
      >
        $ 
        100
         (USD)
      </h3>
      <h3
        class="doge-price"
      >
        1
         (Bitcoin)
      </h3>
    </section>
  </section>
</div>
`;

exports[`DogeTracker Component when \`useFetch\` returns data when 500 status returned displays error messaging 1`] = `
<div>
  <section>
    <h3>
      Whoops! There was a problem loading
       
      Dogecoin price tracker data.
    </h3>
    <p>
      Please reload browser to try again in a little bit.
    </p>
  </section>
</div>
`;
